src/support.js

```diff
@@ -599,48 +599,122 @@ Module['registerFunctions'] = registerFunctions;
 #endif // RELOCATABLE
 #endif // EMULATED_FUNCTION_POINTERS
 
-#if EMULATED_FUNCTION_POINTERS == 0
-#if WASM_BACKEND && RESERVED_FUNCTION_POINTERS
-var jsCallStartIndex = {{{ JSCALL_START_INDEX }}};
-var jsCallSigOrder = {{{ JSON.stringify(JSCALL_SIG_ORDER) }}};
-var jsCallNumSigs = Object.keys(jsCallSigOrder).length;
-var functionPointers = new Array(jsCallNumSigs * {{{ RESERVED_FUNCTION_POINTERS }}});
-#else // WASM_BACKEND && RESERVED_FUNCTION_POINTERS == 0
+#if !WASM_BACKEND && EMULATED_FUNCTION_POINTERS == 0
 var jsCallStartIndex = 1;
 var functionPointers = new Array({{{ RESERVED_FUNCTION_POINTERS }}});
-#endif // WASM_BACKEND && RESERVED_FUNCTION_POINTERS
-#endif // EMULATED_FUNCTION_POINTERS == 0
+#endif // !WASM_BACKEND && EMULATED_FUNCTION_POINTERS == 0
 
 #if WASM
+// Wraps a JS function as a wasm function with a given signature.
+// In the future, we may get a WebAssembly.Function constructor. Until then,
+// we create a wasm module that takes the JS function as an import with a given
+// signature, and re-exports that as a wasm function.
+function convertJsFunctionToWasm(func, sig) {
```

Comment:
Had a bit of fun doing golfing this down... :-)

```js
let w=(f,s)=>{var[R,...P]=s,F=x=>[x.length,...x.map(y=>({i:127,j:126,f:125,d:124
}[y]))],Q=[1,96,...F(P),...F(R=='v'?[]:[R])],W=WebAssembly;return(new W.Instance
(new W.Module(new Uint8Array([0,97,115,109,1,0,0,0,1,Q.length,...Q,2,6,1,0,1,102
,0,0,7,5,1,1,102,0,0])),{'':{f}})).exports.f}

let convertJsFunctionToWasm = w;

const add = (x, y) => x + y;
const addi = convertJsFunctionToWasm(add, 'iii');
const addd = convertJsFunctionToWasm(add, 'ddd');
const printi = convertJsFunctionToWasm(print, 'vi');

print(addi(1.2, 2.3)); // 3
print(addd(1.2, 2.3)); // 3.5
printi(42.1);          // 42
```

Reply:
If the size of the generated module also reduced?  Are you suggesting we use this version?

Reply:
No, sorry, we definitely shouldn't use this code. :smile: just had some insomnia last night so I started trying to see how small I could make it, thought you might appreciate. Feel free to ignore, haha

Reply:
Perhaps the most useful wasm module ever on per-byte basis?

Reply:
Devil's advocate, *should* we not use this code? JS optimizers are extremely unlikely to shrink this code as well, and code size wins are code size wins.

Though it should be `#if`ed to only be shipped in release builds, and if we ever change the non-golfed version we should delete it. That is some write-only code.

Reply:
Probably best not to use it :) Does it even work with emscripten since it uses ES6 features? Was the old uglify pass ever removed?

Reply:
We do support ES6 now in all the major JS passes (but we didn't remove the old ones because they let us support source maps in asm.js), and can probably assume JS is modern in places where wasm is used (but I'm not sure about that).